### PR TITLE
eclass/distutils-r1: add distutils_enable_mkdocs

### PR DESCRIPTION
**Pull request by AndrewAmmerlaan**
based on distutils_enable_sphinx, allows easy enabling
of doc building with mkdocs similar to sphinx

see also: https://github.com/gentoo/gentoo/pull/15150

Package-Manager: Portage-2.3.96, Repoman-2.3.22
Signed-off-by: Andrew Ammerlaan <andrewammerlaan@riseup.net>

### CODE REVIEW

**Comment by mgorny** [comment]
How many packages actually use mkdocs? I see only one.

**Comment by AndrewAmmerlaan** [comment]
> How many packages actually use mkdocs? I see only one.

I know of mkdocs-bootstrap and mkdocs-bootswatch in ::gentoo (https://github.com/gentoo/gentoo/pull/15150)

And:
- dev-games/surgescript @lanodan 
- dev-python/httpx
- dev-python/starlette @oczkers 
- dev-python/uvicorn 
- dev-python/databases @oczkers 
- dev-python/backrefs @Alessandro-Barbieri 
- dev-python/bracex @Alessandro-Barbieri 
- dev-pyhton/pymdown-extensions
- dev-python/wcmatch @Alessandro-Barbieri 
- dev-pyhton/pyspelling @Alessandro-Barbieri 
in ::guru

Not that many packages to be honest, but I imagine there are some other packages that currently do not have USE="doc" but actually do have docs that can be built with mkdocs (e.g.  mkdocs-bootstrap and mkdocs-bootswatch, which didn't have support for building their docs until I added it in the mentioned PR)

The fact that mkdocs-material is missing from ::gentoo is also not helping, because it is apparently a very popular theme. And docs that require it will not build without it. According to Github 2.3k packages use mkdocs-material, which is 39% of the total packages using mkdocs (5.9k according to Github). At the very least, I saw one package in the repo that didn't have USE="doc" support because the docs required mkdocs-material, but I don't recall which one it was. ( mkdocs-material is currently in ::guru, but I might make a PR for that one later, as it is quite a popular theme )

According to Github there are 5.9k packages using mkdocs, which admittedly is not as much as sphinx's 80.2k (an order of magnitude bigger), but it is still 5.9k.  This means that, statistically, I would expect that that for every 14 packages that currently use `distutils_enable_sphinx`, there would be 1 package that would benefit from a 'distutils_enable_mkdocs` function (that would be 7%). 
I don't know exactly how many packages there are that use sphinx, but I imagine there are quite a lot, making the expected number of packages using mkdocs also significant, at the very least more then 1. The fact that we do not observe that many packages that depend on mkdocs, can be explained partly by the lack of an easy function to enable it (which this PR fixes), and the lack of the mkdocs-material theme (which 39% of the mkdocs packages use). 

Furthermore, this is a very easy fix, to having to do this in every ebuild that uses mkdocs:
```
IUSE="doc"

....

BDEPEND="doc? (
         dev-python/mkdocs
         dev-python/mkdocs-material
)"

....

python_compile_all() {
	default
	if use doc; then
		mkdocs build || die "Failed to make docs"
		HTML_DOCS="site"
               rm site/sitempap.xml.gz || die
	fi
}

```

So it might not be that many packages that actually use mkdocs at the moment. But, I've had to put the above lines in quite some ebuilds now, and I felt that a `distutils_enable_mkdocs` function might be useful to clean this up a bit, and not have to do the same over and over again. Especially since it was very easy to add this function using (the very useful) `distutils_enable_sphinx` as a template.

In short, I don't see any disadvantage to adding this function, and it has the potential of being quite useful, at least for 12 packages listed above. And probably for many more packages, both in ::guru and :;gentoo.

**Comment by mgorny** [comment]
I'm wondering how to do this without repeating all the code for every doc builder out there.

**Comment by AndrewAmmerlaan** [comment]
> I'm wondering how to do this without repeating all the code for every doc builder out there.

I've thought about doing something like: `distutils_enable_docs <doc builder> <dir> <deps>` (which would be the ideal way to do it), and having things like `IUSE+="doc", HTML_DOCS+="....` handled by this master function, which then calls the specific `<doc-builder>-build` function. But I didn't want to completely break open the `distutils_enable_sphinx` function. Because then all the ebuilds that now call `distutils_enable_sphinx` would be broken. Should I try to do it with a generic function anyway? (and then revbump the eclass?)

**Comment by lanodan** [comment]
Are you sure that this one should be into python eclasses? mkdocs is pretty much language independent (we use it in an elixir+vuejs project).

And yeah, I think mkdocs is or will be used a lot in projects in general, as it's a quite good documentation generator that often works better than the language-provided one or stuff like doxygen.

**Comment by AndrewAmmerlaan** [comment]
> Are you sure that this one should be into python eclasses? mkdocs is pretty much language independent (we use it in an elixir+vuejs project).

No, I'm not sure about that at all. I put it in distutils-r1 because that is where the sphinx function is, and even though mkdocs supports more then just Python, it does look like most of the projects using mkdocs are Python projects. I suppose I could put it in a separate eclass, but then it wouldn't be possible to combine it with the sphinx function in a generic `enable_docs <doc builder>` function. 

Or maybe we could have this separate eclass (`docs.eclass`?), support Sphinx as well? Doesn't Sphinx also support more then just Python? This way it might also be easier to add even more doc builders without having to spam the distutils-r1 eclass with them. @mgorny what do you think of a separate `docs.eclass`?

**Comment by AndrewAmmerlaan** [comment]
> I'm wondering how to do this without repeating all the code for every doc builder out there.

> Are you sure that this one should be into python eclasses? mkdocs is pretty much language independent (we use it in an elixir+vuejs project).

@mgorny @lanodan I just added a commit here containing a first draft for a `docs.eclass`. It adds a couple of variables:
`DOCBUILDER=<sphinx | mkdocs>` Builder to use
`DOCDIR=<dir>` Directory mkdocs.yml/sphinx.conf is in (defaults to ${S})
`DOCDEPEND=<deps>` Additional documentation dependencies to use
`AUTODOC=<0|1>` Controls the use of sphinx.ext.autodoc and mkautodoc

I tried my best to write it as generic as possible, and to make it relatively easy to add more doc builders. The eclass was producing the expected results when I tested it just now on some ebulds in ::guru (both mkdocs/sphinx with/without AUTODOC). Please let me know what you think of the eclass, there is probably still room for improvement.

**Comment by Alessandro-Barbieri** [comment]
I like the idea of the docs eclass

**Comment by AndrewAmmerlaan** [comment]
>@mgorny @lanodan I just added a commit here containing a first draft for a docs.eclass. It adds a couple of variables:
DOCBUILDER=<sphinx | mkdocs> Builder to use
DOCDIR=<dir> Directory mkdocs.yml/sphinx.conf is in (defaults to ${S})
DOCDEPEND=<deps> Additional documentation dependencies to use
AUTODOC=<0|1> Controls the use of sphinx.ext.autodoc and mkautodoc

> Are you sure that this one should be into python eclasses? mkdocs is pretty much language independent (we use it in an elixir+vuejs project).

> I like the idea of the docs eclass

I pushed my docs.eclass to ::guru, after the fixes I committed today it should be in a functioning state. Please feel free to test it on the packages in ::guru that require mkdocs (or sphinx and can't use distutils for whatever reason). Let me know if you encounter any issues or if the in-file documentation is unclear.
